buffer/src/main/java/io/netty5/buffer/api/bytebuffer/NioBuffer.java

```diff
@@ -1137,23 +1148,30 @@ private RuntimeException checkWriteState(IndexOutOfBoundsException ioobe, int of
 
     private RuntimeException readAccessCheckException(int index, int size) {
         if (rmem == CLOSED_BUFFER) {
-            throw bufferIsClosed(this);
+            return bufferIsClosed(this);
         }
         return outOfBounds(index, size);
     }
 
-    private RuntimeException writeAccessCheckException(int index, int size) {
+    private void writeAccessCheckException(int index, int size, boolean mayExpand) {
         if (rmem == CLOSED_BUFFER) {
             throw bufferIsClosed(this);
         }
         if (wmem != rmem) {
-            return bufferIsReadOnly(this);
+            throw bufferIsReadOnly(this);
         }
-        return outOfBounds(index, size);
+        int capacity = capacity();
+        if (mayExpand && index > 0 && index <= capacity && capacity < 131072 /* 128k */ && isOwned()) {
+            int minimumGrowth = PlatformDependent.roundToPowerOfTwo(capacity * 2) - capacity; // Grow into power-of-two.
+            ensureWritable(size, minimumGrowth, false);
+            checkSet(index, size); // Verify writing is now possible, without recursing.
+            return;
+        }
+        throw outOfBounds(index, size);
     }
 
     private IndexOutOfBoundsException outOfBounds(int index, int size) {
```

Comment:
nit: consider renaming to something like `checkOutOfBounds(...)` as it is throwing now directly

Reply:
This one should go back to returning the exception, since there's at least one place where we want to add a suppressed exception.